src/hotspot/share/logging/logFileOutput.cpp

```diff
@@ -304,12 +312,28 @@ int LogFileOutput::write(const LogDecorations& decorations, const char* msg) {
   return written;
 }
 
+int LogFileOutput::write(const LogDecorations& decorations, const char* msg) {
+  if (_stream == NULL) {
+    // An error has occurred with this output, avoid writing to it.
+    return 0;
+  }
+
+  LogAsyncFlusher* flusher = LogAsyncFlusher::instance();
+  if (_async_mode && flusher != NULL) {
```

Comment:
hi, @simonis  and @tstuefe 

The purpose of nullcheck here is to circumvent the problem of circular dependency which also mentioned by Thomas.  Logging subsystem as a low-level infrastructure could be invoked as early as `LogConfiguration::initialize` has done in `Threads::create_vm()`. At this time, mutex and Thread haven't been initialized, therefore, I can't initialize `LogAsyncFlusher` appropriately. 

The reason I provide separated APIs `LogAsyncFlusher::initialize()` & `LogAsyncFlusher::instance()` because I need to separate log subsystems into 2 phrases. LogAsyncFlusher::initialize() marks an important point. Before this point, logging subsystem can only use synchronous mode no matter what. After this point, async_mode takes effect because buffer has been established.  That is to say, all code in this PR can only take over logging after that point.  

I read Thomas concerned about fancy calls such as strdup, os::malloc etc.  Instead of designing async-logging absolutely self-contained and low-level, I chose a common technique called 'lazy-initialization'.  I don't think async mode is absolutely necessary in the startup time of hotspot. The gain is obvious. I can use hotspot's established constructs to simplify the implementation. 

May I ask you and Thomas to comment on this design? If it works, I will add more comments about it.